Make bold synthesis less agressive

vitorroriz opened the pull request:
#### 0bf8c7a4ae3fb59b0cca0c79b8f97a54a8914cac
<pre>
Make bold synthesis less agressive
<a href="https://bugs.webkit.org/show_bug.cgi?id=284340">https://bugs.webkit.org/show_bug.cgi?id=284340</a>
<a href="https://rdar.apple.com/138047199">rdar://138047199</a>

Reviewed by Brent Fulgham and Tim Nguyen.

We currently synthesize &apos;bold&apos; by expanding a glyph&apos;s bounding
box by a certain offset. On macOS this offset is always fixed
and don&apos;t scale with the font size. This makes the font to appear
signficantly more bold than other engines for reduced font sizes.

We are now using an offset that scales with the font size for macOS.
We already do that for iOS, so we are unifying both paths and tweaking
the scaling factor.

* LayoutTests/fast/css3-text/font-synthesis-bold-small-font-size-expected.html: Added.
* LayoutTests/fast/css3-text/font-synthesis-bold-small-font-size.html: Added.
* Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
(WebCore::Font::platformInit):

Canonical link: <a href="https://commits.webkit.org/288130@main">https://commits.webkit.org/288130@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0c44ede5fe27773efccf211783cb265d6e2979cf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81928 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1455 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35885 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86485 "") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32959 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84034 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1490 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9278 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/86485 "") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21612 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84998 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1101 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74559 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/86485 "") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1003 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28736 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31379 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72336 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29345 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87916 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9170 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6533 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72254 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9355 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70378 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71476 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17819 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15580 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14505 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/556 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9121 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14657 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8961 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12486 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10769 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rreno commented on Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
```diff
@@ -117,11 +117,8 @@ bool fontHasEitherTable(CTFontRef ctFont, unsigned tableTag1, unsigned tableTag2
-#else
-    m_syntheticBoldOffset = m_platformData.syntheticBold() ? 1.0f : 0.f;
-#endif
+    auto constexpr syntheticBoldScaleFactor = 36.0f;
```
Could you write something in the commit message to explain where this number comes from? Even if it's just "I chose a number that looked nice" it at least leaves a paper trail for the magic constant.

brentfulgham commented on LayoutTests/fast/css3-text/font-synthesis-bold-small-font-size-expected.html:
```diff
@@ -0,0 +1,22 @@
+</style>
+Test passes if block occupies only 1 line:
+<div class="test">a</div>
+</html>
```
Could this be a WPT instead? (@nt1m, what do you think?)

nt1m replied:
The exact amount of boldness for synthesis isn't specified, so it would be better to not make a WPT.

vitorroriz committed:
Make bold synthesis less agressive
https://bugs.webkit.org/show_bug.cgi?id=284340
rdar://138047199

Reviewed by Brent Fulgham and Tim Nguyen.

We currently synthesize 'bold' by expanding a glyph's bounding
box by a certain offset. On macOS this offset is always fixed
and don't scale with the font size. This makes the font to appear
signficantly more bold than other engines for reduced font sizes.

We are now using an offset that scales with the font size for macOS.
We already do that for iOS, so we are unifying both paths and tweaking
the scaling factor.

* LayoutTests/fast/css3-text/font-synthesis-bold-small-font-size-expected.html: Added.
* LayoutTests/fast/css3-text/font-synthesis-bold-small-font-size.html: Added.
* Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
(WebCore::Font::platformInit):

Canonical link: https://commits.webkit.org/288130@main
